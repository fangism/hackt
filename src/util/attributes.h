/**
	\file "util/attributes.h"
	Configure-dependent use of compiler attributes.  
	This file is highly dependent on compiler characteristics.  
	TODO: define these conditionally depending on configure.
	Hint: use AC_COMPILE_IFELSE in configure.ac
	$Id: attributes.h,v 1.7 2006/02/25 04:55:03 fang Exp $
 */

#ifndef	__UTIL_ATTRIBUTES_H__
#define	__UTIL_ATTRIBUTES_H__

#include "config.h"
/* use configure to determine whether or not compiler supports attributes */

/*===========================================================================*/
/* general attributes */
/**
	Apply to types that are never used, but do some real work
	in their non-trivial constructors and destructors.  
	e.g. stacktrace, index, thread_lock, etc...
	Also use for variable that are only used in assertion checks, 
	which are disabled by -DNDEBUG
	If all else fails, you can always -Wno-unused or -Wno-error
 */
#if defined(HAVE_ATTRIBUTE_UNUSED)
#define	__ATTRIBUTE_UNUSED__		__attribute__ ((unused))
#else
#define	__ATTRIBUTE_UNUSED__
#endif

/**
	In constructor syntax, some compilers like the attribute before the
	constructor arguments, some after.  YMMV.
	e.g.	foo bar __attribute__((unused)) (x, y, z);
	e.g.	foo bar (x, y, z) __attribute__((unused));
	To handle these cases automatically, wrap around the
	constructor arguments like so (yes, double parens are needed):
	foo bar __ATTRIBUTE_UNUSED_CTOR__((x, y, z));
 */
#ifdef HAVE_ATTRIBUTE_UNUSED
#ifdef HAVE_ATTRIBUTE_UNUSED_BEFORE_CTOR
	/* known: g++-4.0 */
#define	__ATTRIBUTE_UNUSED_CTOR__(args)		__ATTRIBUTE_UNUSED__ args
#elif defined(HAVE_ATTRIBUTE_UNUSED_AFTER_CTOR)
	/* known: g++-3.3 */
#define	__ATTRIBUTE_UNUSED_CTOR__(args)		args __ATTRIBUTE_UNUSED__
#else
/* #error I don't know where __attribute__((unused)) belongs! */
#define	__ATTRIBUTE_UNUSED_CTOR__(args)		args
#endif
#else
#define	__ATTRIBUTE_UNUSED_CTOR__(args)		args
#endif

/*===========================================================================*/
/* visibility attributes (support added in gcc-4.0) */

/**
	Public symbol visibility.  
 */
#ifdef HAVE_ATTRIBUTE_VISIBILITY_DEFAULT
#define	__VISIBILITY_DEFAULT__	__attribute__ ((visibility("default")))
#else
#define	__VISIBILITY_DEFAULT__
#endif

/**
	Hidden symbol visibility.  
 */
#ifdef HAVE_ATTRIBUTE_VISIBILITY_HIDDEN
#define	__VISIBILITY_HIDDEN__	__attribute__ ((visibility("hidden")))
#else
#define	__VISIBILITY_HIDDEN__
#endif

/*===========================================================================*/
/* type and variable attributes */

/**
	Align to natural boundary.  
 */
#ifdef	HAVE_ATTRIBUTE_ALIGNED
#define	__ATTRIBUTE_ALIGNED__		__attribute__ ((aligned))
#else
#define	__ATTRIBUTE_ALIGNED__
#endif

/**
	Align to manually sized boundary.  
 */
#ifdef	HAVE_ATTRIBUTE_ALIGNED_SIZE
#define	__ATTRIBUTE_ALIGNED_SIZE__(x)	__attribute__ ((aligned(x)))
#else
#define	__ATTRIBUTE_ALIGNED_SIZE__(x)
#endif

/**
	Pack structure into minimal amount of space regardless of 
	alignment requirements.  
 */
#ifdef	HAVE_ATTRIBUTE_PACKED
#define	__ATTRIBUTE_PACKED__		__attribute__ ((packed))
#else
#define	__ATTRIBUTE_PACKED__
#endif

/*===========================================================================*/
/* function attributes */

/**
	Use this in functions that do not return, such as exit and abort
	style functions.  
 */
#ifdef HAVE_ATTRIBUTE_NORETURN
#define	__ATTRIBUTE_NORETURN__		__attribute__ ((noreturn))
#else
#define	__ATTRIBUTE_NORETURN__
#endif

/*- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -*/
/**
	The const attribute asserts that this function has no
	side-effects: it only takes arguments and produces a 
	return value.  
 */
#ifdef HAVE_ATTRIBUTE_CONST
#define	__ATTRIBUTE_CONST__		__attribute__ ((const))
#else
#define	__ATTRIBUTE_CONST__
#endif

#ifdef HAVE_ATTRIBUTE_PURE
#define	__ATTRIBUTE_PURE__		__attribute__ ((pure))
#else
#define	__ATTRIBUTE_PURE__
#endif

/*- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -*/
/**
	The malloc attribute asserts that this non-NULL pointers returned
	by this function do not alias, just like malloc.
	This helps alias-analysis based optimizations.  
 */
#ifdef HAVE_ATTRIBUTE_MALLOC
#define	__ATTRIBUTE_MALLOC__		__attribute__ ((malloc))
#else
#define	__ATTRIBUTE_MALLOC__
#endif


/*===========================================================================*/

#endif	/* __UTIL_ATTRIBUTES_H__ */

