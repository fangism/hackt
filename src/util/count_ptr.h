// "count_ptr.h"
// reference counting pointer class

/**
	Generally not intended for mixed use with pointers from
	"ptrs.h": excl_ptr, some_ptr, never_ptr.  

	proposals:
		Allow release of naked pointer when there is only 
		one sole reference remaining, otherwise return NULL.
		Will make it interfaceable with excl_ptr, and some_ptr.  
**/

#ifndef	__COUNT_PTR_H__
#define	__COUNT_PTR_H__

#include <assert.h>
#include "count_ptr_fwd.h"

#ifndef NULL
#define	NULL		0
#endif

/**
	Namespace for reference-count pointer-classes.  
 */
namespace COUNT_PTR_NAMESPACE {
//=============================================================================
/**
	Base class for all reference-count pointers.  
 */
class abstract_base_count_ptr {
	// parent could be abstract_ptr
template <class> friend class count_ptr;
template <class> friend class count_const_ptr;

protected:
	/**
		The reference count.  Need not be mutable.  
	 */
	size_t*	ref_count;

protected:
	abstract_base_count_ptr(size_t* s = NULL) : ref_count(s) { }

	/**
		This destructor does absolutely nothing.  
		It is subclasses' responsibility to take care of 
		deleting ref_count when appropriate.  
	 */
virtual	~abstract_base_count_ptr() { }

	/**
		Raw pointer.  
	 */
virtual	void* void_ptr(void) const = 0;

	/**
		Raw pointer, const.  
	 */
virtual	const void* const_void_ptr(void) const = 0;

public:
	/**
		Reference count.  
	 */
	size_t refs(void) const {
		if (ref_count) return *ref_count;
		else return 0;
	}

	/**
		A shared reference is not really owned by any one
		copy of the pointer.  
		The last reference to an object is the owner.  
	 */
	bool owned(void) const {
		return (ref_count && *ref_count == 1);
	}

	/**
		Ditches pointer without returning it, 
		maintaining reference count, deleting when appropriate.  
	 */
virtual	void abandon(void) = 0;
	/**
		Alias for abandon.
	 */
	void nullify(void) { abandon(); }
	void ditch(void) { abandon(); }
	void scrap(void) { abandon(); }

	/**
		Test for nullity.  
		Will overriding in sub-classes be faster, 
		without virtual call?
	 */
	operator bool () const { return const_void_ptr() != NULL; }

	bool operator == (const void* p) const
		{ return const_void_ptr() == p; }
	bool operator != (const void* p) const
		{ return const_void_ptr() != p; }
	bool operator == (const abstract_base_count_ptr& p) const
		{ return const_void_ptr() == p.const_void_ptr(); }
	bool operator != (const abstract_base_count_ptr& p) const
		{ return const_void_ptr() != p.const_void_ptr(); }

};	// end class abstract_base_count_ptr

//=============================================================================
/**
	Reference counted pointer, non-const flavor.  
	How to use...
 */
template <class T>
class count_ptr : public abstract_base_count_ptr {
template <class> friend class count_ptr;
template <class> friend class count_const_ptr;
protected:
	typedef	abstract_base_count_ptr		parent;

protected:
	/**
		The pointer to the referenced object.  
	 */
	T*	ptr;

	void* void_ptr(void) const { return ptr; }
	const void* const_void_ptr(void) const { return ptr; }

public:
	/**
		Constructor of reference-counter pointer.  
		Stupid things to avoid: 
		1) tampering with (leaking out) newly allocated
			pointer before passing to this constructor.  
		\param p should be a newly allocated pointer, ONLY.  
	 */
explicit count_ptr(T* p = NULL) : parent((p) ? new size_t(1) : NULL), ptr(p) { }

	/**
		Copy constructor.  
		No argument pointer should ever be NULL?
	 */
	count_ptr(const count_ptr<T>& c) : parent(c.ref_count), ptr(c.ptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}

	/**
		Safe, up-cast copy constructor.  
		Still maintains reference count.  
	 */
	template <class S>
	count_ptr(const count_ptr<S>& c) : parent(c.ref_count), ptr(c.ptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}

	/**
		UNSAFE: constructor that corecively sets the 
		reference pointer and the count pointer 
		and increments the pointer.  
		The manager should initially allocate a counter set to zero.  
		The manager should not own the pointer or the counter.  
		Needed for transient pointer to object reconstruction.  
		Use this only if you know what you're doing.
	 */
	count_ptr(T* p, size_t* c) : parent(c), ptr(p) {
		if (p) {
			assert(c);		// counter must be valid
			(*c)++;			// increment here
		} else
			assert(!c);		// counter must also be NULL
	}

protected:
	/**
		Overkill safety checks.  
	 */
	T* release(void) {
		T* ret = ptr;
		if(this->ref_count) {
			(*this->ref_count)--;
			if (!*this->ref_count) {
				assert(ptr);
				delete ptr;
				delete this->ref_count;
			}
			ptr = NULL;
			this->ref_count = NULL;
		} else {
			assert(!ptr);
		}
		return ret;
	}

	/**
		\param p the new pointer, from some other count_ptr.
		\param c the corresponding reference count if p is valid, 
			else is ignored.  
	 */
	void reset(T* p, size_t* c) {
		if(this->ref_count) {
			(*this->ref_count)--;
			if (!*this->ref_count) {
				assert(ptr);
				delete ptr;
				delete this->ref_count;
			}
		} else {
			assert(!ptr);
		}
		ptr = p;
		this->ref_count = (ptr) ? c : NULL;
		if (ptr && this->ref_count)
			(*this->ref_count)++;
	}

public:
	/**
		Destructor, which conditionally deletes pointer if is
		last reference.  
	 */
	~count_ptr() { release(); }

	void abandon(void) { release(); }

	/**
		Dereference.  
	 */
	T& operator * (void) const throw() { assert(ptr); return *ptr; }

	/**
		Dereference to member or method.  
	 */
	T* operator -> (void) const throw() { assert(ptr); return ptr; }

	/**
		Assignment operator.  
	 */
	count_ptr<T>& operator = (const count_ptr<T>& c) {
		reset(c.ptr, c.ref_count);
		return *this;
	}

	/**
		Returns a naked pointer only if this is the final reference
		to the object, otherwise returns NULL.  
	 */
	T* exclusive_release(void) {
		if (this->owned()) {
			T* ret = ptr;
			ptr = NULL;
			delete this->ref_count;
			this->ref_count = NULL;
			return ret;
		} else
			return NULL;
	}

	/**
		Static cast.  
	 */
	template <class S>
	count_ptr<S> as_a(void) const {
		count_ptr<S> ret;
		ret.reset(static_cast<S*>(this->ptr), this->ref_count);
		return ret;
	}

	/**
		Dynamic cast.  
	 */
	template <class S>
	count_ptr<S> is_a(void) const {
		count_ptr<S> ret;
		ret.reset(dynamic_cast<S*>(this->ptr), this->ref_count);
		return ret;
	}

	/**
		Dynamic cast assertion.  
	 */
	template <class S>
	void must_be_a(void) const {
		assert(dynamic_cast<S*>(this->ptr));
	}

};	// end class count_ptr

//=============================================================================
/**
	Read-only version of reference-counted pointer.  
 */
template <class T>
class count_const_ptr : public abstract_base_count_ptr {
template <class> friend class count_const_ptr;

protected:
	typedef	abstract_base_count_ptr		parent;
protected:
	const T*	cptr;

	void* void_ptr(void) const { return NULL; }
	const void* const_void_ptr(void) const { return cptr; }

public:

	/**
		Constructor of reference-counter pointer.  
		Stupid things to avoid: 
		1) tampering with (leaking out) newly allocated
			pointer before passing to this constructor.  
		\param p should be a newly allocated pointer, ONLY.  
	 */
explicit count_const_ptr(const T* p = NULL) :
		parent((p) ? new size_t(1) : NULL), cptr(p) { }

	/**
		Copy constructor.  
		No argument pointer should ever be NULL?
	 */
	count_const_ptr(const count_ptr<T>& c) :
		parent(c.ref_count), cptr(c.ptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}
	count_const_ptr(const count_const_ptr<T>& c) :
		parent(c.ref_count), cptr(c.cptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}

	/**
		Safe, up-cast copy constructor.  
		Still maintains reference count.  
	 */
	template <class S>
	count_const_ptr(const count_ptr<S>& c) :
		parent(c.ref_count), cptr(c.ptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}
	template <class S>
	count_const_ptr(const count_const_ptr<S>& c) :
		parent(c.ref_count), cptr(c.cptr) {
		if (this->ref_count)
			(*this->ref_count)++;
	}

	/**
		UNSAFE: constructor that corecively sets the 
		reference pointer and the count pointer 
		and increments the pointer.  
		The manager should initially allocate a counter set to zero.  
		The manager should not own the pointer or the counter.  
		Needed for transient pointer to object reconstruction.  
		Use this only if you know what you're doing.
	 */
	count_const_ptr(const T* p, size_t* c) : parent(c), cptr(p) {
		if (p) {
			assert(c);		// counter must be valid
			(*c)++;			// increment here
		} else
			assert(!c);		// counter must also be NULL
	}

protected:
	/**
		Overkill safety checks.  
	 */
	const T* release(void) {
		const T* ret = cptr;
		if(this->ref_count) {
			(*this->ref_count)--;
			if (!*this->ref_count) {
				assert(cptr);
				delete cptr;
				delete this->ref_count;
			}
			cptr = NULL;
			this->ref_count = NULL;
		} else {
			assert(!cptr);
		}
		return ret;
	}

	/**
		\param p the new pointer, from some other count_const_ptr.
		\param c the corresponding reference count if p is valid, 
			else is ignored.  
	 */
	void reset(const T* p, size_t* c) {
		if(this->ref_count) {
			(*this->ref_count)--;
			if (!*this->ref_count) {
				assert(cptr);
				delete cptr;
				delete this->ref_count;
			}
		} else {
			assert(!cptr);
		}
		cptr = p;
		this->ref_count = (cptr) ? c : NULL;
		if (cptr && this->ref_count)
			(*this->ref_count)++;
	}

public:
	/**
		Destructor, which conditionally deletes pointer if is
		last reference.  
	 */
	~count_const_ptr() { release(); }

	void abandon(void) { release(); }

	/**
		Dereference.  
	 */
	const T& operator * (void) const throw() { assert(cptr); return *cptr; }

	/**
		Dereference to member or method.  
	 */
	const T* operator -> (void) const throw() { assert(cptr); return cptr; }

	/**
		Assignment operator.  
	 */
	count_const_ptr<T>& operator = (const count_ptr<T>& c) {
		reset(c.ptr, c.ref_count);
		return *this;
	}
	count_const_ptr<T>& operator = (const count_const_ptr<T>& c) {
		reset(c.cptr, c.ref_count);
		return *this;
	}

	/**
		Returns a naked pointer only if this is the final reference
		to the object, otherwise returns NULL.  
	 */
	const T* exclusive_release(void) {
		if (this->owned()) {
			const T* ret = cptr;
			cptr = NULL;
			delete this->ref_count;
			this->ref_count = NULL;
			return ret;
		} else
			return NULL;
	}

	/**
		Static cast.  
	 */
	template <class S>
	count_const_ptr<S> as_a(void) const {
		count_const_ptr<S> ret;
		ret.reset(static_cast<const S*>(this->cptr), this->ref_count);
		return ret;
	}

	/**
		Dynamic cast.  
	 */
	template <class S>
	count_const_ptr<S> is_a(void) const {
		count_const_ptr<S> ret;
		ret.reset(dynamic_cast<const S*>(this->cptr), this->ref_count);
		return ret;
	}

	/**
		Dynamic cast assertion.  
	 */
	template <class S>
	void must_be_a(void) const {
		assert(dynamic_cast<const S*>(this->cptr));
	}

};	// end class count_const_ptr

//=============================================================================
}	// end namespace COUNT_PTR_NAMESPACE

#endif	//	__COUNT_PTR_H__

