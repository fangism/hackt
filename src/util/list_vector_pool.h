/**
	\file "list_vector_pool.h"
	Simple template container-based memory pool.  
	Basically allocates a large chunk at a time.  

	$Id: list_vector_pool.h,v 1.1 2004/11/05 02:38:47 fang Exp $
 */

#ifndef	__LIST_VECTOR_POOL_H__
#define	__LIST_VECTOR_POOL_H__

#include <pthread.h>

#include <assert.h>
#include <queue>
#include <list>
#include <vector>


// because FreeBSD's <pthreads.h> defines initializer as NULL
// and I haven't figured out a workaround yet.
#if defined(__FreeBSD__)
// #warn	"This allocator will not be thread-safe in FreeBSD... yet."
#define	THREADED_ALLOC		0
#else
#define	THREADED_ALLOC		1
#endif

// turn invariant assertions on or off
#define	DEBUG_LIST_VECTOR_POOL	1

// annoying debug messages
#define VERBOSE_ALLOC		1 && DEBUG_LIST_VECTOR_POOL
#define	VERBOSE_LOCK		0 && VERBOSE_ALLOC

#if VERBOSE_ALLOC
#include <iostream>
#endif

#if DEBUG_LIST_VECTOR_POOL
#define INVARIANT_ASSERT(foo)	assert(foo)
#else
#define INVARIANT_ASSERT(foo)	{ }
#endif

//=============================================================================
/**
	Memory-management utility namespace.
	Should implement the std::allocator interface.  
 */
namespace memory {
typedef	pthread_mutex_t				mutex_type;
using std::queue;
using std::list;
using std::vector;

#if VERBOSE_ALLOC
using namespace std;		// for debugging
#endif

//=============================================================================
// forward declaration
template <class T, bool Threaded = true>
class list_vector_pool;

//-----------------------------------------------------------------------------
// specialization
template <>
class list_vector_pool<void> {
public:
	typedef size_t		size_type;
	typedef ptrdiff_t	difference_type;
	typedef void*		pointer;
	typedef const void*	const_pointer;
	typedef void		value_type;
// no rebind yet...
};

//-----------------------------------------------------------------------------
#if 0
// specialization for when allocated type is just pointer, 
// to prevent pointer recursion, not needed unless free_list_type is fancy
template <class T, bool Threaded>
class list_vector_pool<T*, Threaded> {
public:
	typedef	T*				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	value_type*			pointer;
	typedef	const value_type*		const_pointer;
	typedef	value_type&			reference;
	typedef	const value_type&		const_reference;
private:
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
	// plain queue to stop pointer recursion
	typedef	queue<pointer>			free_list_type;
public:

};	// end class list_vector_pool specialization
#endif

//-----------------------------------------------------------------------------
/**
	Simple, efficient memory pool.  
	No alignment or placement.  
	Only good for allocating one element at a time, 
	so is not general purpose.  
	For a more general pool allocator use the __gnu_cxx::__pool_alloc, 
	which is likely to be the default allocator.  
 */
template <class T, bool Threaded>
class list_vector_pool {
public:
	typedef	T				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	T*				pointer;
	typedef	const T*			const_pointer;
	typedef	T&				reference;
	typedef	const T&			const_reference;
private:
	// could propagate thread-sensitivity parameter to the respective
	// STL data-structures' allocators...
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
#if 1
	typedef	queue<pointer>			free_list_type;
#else
	// ballsy: use itself as allocator! (template recursion problem?)
	typedef	queue<pointer, deque<pointer, list_vector_pool<pointer> > >
						free_list_type;
#endif
#if THREADED_ALLOC
	static mutex_type			the_mutex;
	friend struct Lock;			// grant access to the_mutex
	struct Lock {
		Lock() {
//			assert(the_mutex);
#if VERBOSE_LOCK
			cerr << "getting lock at " << &the_mutex << "... ";
#endif
			if (Threaded)
				assert(!pthread_mutex_lock(&the_mutex));
#if VERBOSE_LOCK
			cerr << "got it." << endl;
#endif
		}

		~Lock() {
#if VERBOSE_LOCK
			cerr << "releasing lock... ";
#endif
			if (Threaded)
				assert(!pthread_mutex_unlock(&the_mutex));
#if VERBOSE_LOCK
			cerr << "released." << endl;
#endif
		}
	};	// end struct Lock
	// may need __attribute__ ((__unused__))
#endif	// THREADED_ALLOC

private:
	// probably no need to change...
	const size_t				chunk_size;
	impl_type				pool;
	free_list_type				free_list;
public:
	/**
		Reserves one chunk up-front.  
		\param C is number of elements to allocate at a time.
			Should be related to page_size/sizeof(T).  
	 */
	list_vector_pool(const size_type C = 16) : 
			chunk_size(C), pool(), free_list() {
		assert(chunk_size);
		// worry about alignment, placement and pages sizes later
		pool.push_back(chunk_type());
		pool.back().reserve(chunk_size);
		INVARIANT_ASSERT(pool.back().capacity() == chunk_size);
		INVARIANT_ASSERT(!pool.back().size());
		INVARIANT_ASSERT(free_list.empty());
	}

	// no copy-constructor

	// default destructor suffices

	/// allocate one element only, without construction
	pointer
	allocate(void) {
#if THREADED_ALLOC
		// volatile? do not optimize away?
		Lock got_the_mutex;
#endif
		// see if an address is available from the free-list
		if (!free_list.empty()) {
			pointer ret = free_list.front();
			INVARIANT_ASSERT(ret);
			free_list.pop();
#if VERBOSE_ALLOC
			cerr << "alloc from free-list @ " << ret << endl;
#endif
			return ret;
		} else {
			// if not, use the chunk.
			INVARIANT_ASSERT(!pool.empty());
			chunk_type* chunk = &pool.back();
			INVARIANT_ASSERT(chunk->capacity());
			if (chunk->size() == chunk->capacity()) {
				// rare case: allocate chunk
#if VERBOSE_ALLOC
				cerr << "new chunk of " << chunk_size
					<< " allocated." << endl;
#endif
				pool.push_back(chunk_type());
				chunk = &pool.back();
				chunk->reserve(chunk_size);
				INVARIANT_ASSERT(!chunk->size());
			}
			INVARIANT_ASSERT(chunk->capacity() == chunk_size);
			chunk->push_back(T());
			// T must be copy-constructible
			pointer ret = &chunk->back();
			INVARIANT_ASSERT(ret);
#if VERBOSE_ALLOC
			cerr << "alloc from pool @ " << ret << endl;
#endif
			return ret;
		}
		// lock will expire at end-of-scope
	}

	/// allocate n elements without construction, unimplemented
	pointer
	allocate(size_type n, list_vector_pool<void>::const_pointer hint = 0);

	/// deallocates without destruction
	// any safety checks for multiple deletions?
	// if so, use set<pointer> for the free_list
	void
	deallocate(pointer p) {
#if THREADED_ALLOC
		Lock got_the_mutex;
#endif
		assert(p);
#if VERBOSE_ALLOC
		cerr << "returned " << p << " to free-list." << endl;
#endif
		free_list.push(p);
		// lock will expire at end-of-scope
	}

	/// deallocates without destruction, unimplemented
	void
	deallocate(pointer p, size_type n);

	/**
		Initialize *p by val.  
		The new operator is given a placed address, 
		which doesn't actually reallocate.  
		The value is copy-constructed in-place.  
	 */
	void
	construct(pointer p, const T& val) {
		assert(p);
		new(p) T(val);
	}

	/// destroys *p without deallocating
	void
	destroy(pointer p) {
		assert(p);
		p->~T();
	}

	// other miscellaneous methods
	size_type max_size() const;

	// rebinding typedef

};	// end class list_vector_pool

//-----------------------------------------------------------------------------
#if THREADED_ALLOC
// static initializer of STL mutex element
template<class T, bool Threaded>
mutex_type
list_vector_pool<T, Threaded>::the_mutex = PTHREAD_MUTEX_INITIALIZER;

#endif

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
inline
bool
operator == (const list_vector_pool<T>&, const list_vector_pool<T>&) {
	return true;
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
inline
bool
operator != (const list_vector_pool<T>&, const list_vector_pool<T>&) {
	return false;
}

//=============================================================================
}	// end namespace memory


#undef	INVARIANT_ASSERT

#endif	//	__LIST_VECTOR_POOL_H__

