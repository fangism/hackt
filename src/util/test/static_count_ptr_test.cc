/**
	'static_count_ptr_test.cc"
	Testing "unsafe" uses of count_ptr.  
	$Id: static_count_ptr_test.cc,v 1.7 2006/04/27 00:17:58 fang Exp $
 */

#define	ENABLE_STACKTRACE		1

#define	ENABLE_STATIC_TRACE		0

#ifdef	NDEBUG
#undef	NDEBUG
#endif

#include "util/static_trace.hh"
STATIC_TRACE_BEGIN("static_count_ptr_test::main")

#include <iostream>
#include "util/memory/count_ptr.hh"
#include "util/memory/list_vector_pool.hh"
#include "named_pooled_thing.hh"
#include "util/stacktrace.hh"
#include "util/using_ostream.hh"
#include "util/attributes.h"

using namespace util::memory;

typedef	count_ptr<named_thing>	ref_type;

// REQUIRES_STACKTRACE_STATIC_INIT
// required only if dynamically allocating during static initialization
// REQUIRES_LIST_VECTOR_POOL_STATIC_INIT(named_thing)

ref_type
factory_func(void) {
	static named_thing* local_obj = new named_thing;
	STATIC_RC_POOL_REF_INIT;
	static size_t* local_count = NEW_SIZE_T;
	static const size_t zero __ATTRIBUTE_UNUSED__ = (*local_count = 0);
	return ref_type(local_obj, local_count);
}

int
main(int, char*[]) {
	STACKTRACE_VERBOSE;
#if 1
	static const ref_type static_ref __ATTRIBUTE_UNUSED__ = factory_func();

	const ref_type local_ref = factory_func();

	int i = 0;
	for ( ; i < 4; i++) {
		STACKTRACE("for-loop");
		const ref_type for_ref = factory_func();
		cout << "count = " << for_ref.refs() << endl;
	}

	cout << "count = " << local_ref.refs() << endl;
#endif
	return 0;
}

STATIC_TRACE_END("static_count_ptr_test::main")

