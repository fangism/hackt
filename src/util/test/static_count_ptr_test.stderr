START-initialization: named_pooled_thing.o
\-{ list_vector_pool()
Reserved named_thing chunk of size 16*4 starting at ##ADDR##
/-} list_vector_pool()
END---initialization: named_pooled_thing.o
\-{ int main(int, char**)
| \-{ operator new
| | \-{ list_vector_pool::allocate()
| | | \-{ empty ctor
##ADDR##
| | | /-} empty ctor
| | | \-{ dtor
##ADDR## -> 
| | | /-} dtor
Allocated named_thing from pool @ ##ADDR##
| | /-} list_vector_pool::allocate()
| /-} operator new
| \-{ empty ctor
##ADDR##
| /-} empty ctor
| \-{ for-loop
count = 3
| /-} for-loop
| \-{ for-loop
count = 3
| /-} for-loop
| \-{ for-loop
count = 3
| /-} for-loop
| \-{ for-loop
count = 3
| /-} for-loop
count = 2
/-} int main(int, char**)
\-{ dtor
##ADDR## -> 
/-} dtor
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned named_thing @ ##ADDR## to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
START-destruction: named_pooled_thing.o
\-{ ~list_vector_pool()
~list_vector_pool<named_thing>() at ##ADDR##
| \-{ list_vector_pool::status()
	1 chunks of 16*4 named_thing allocated.
	Peak usage: 1 elements, free-list has 1 remaining.
| /-} list_vector_pool::status()
| \-{ empty ctor
##ADDR##
| /-} empty ctor
/-} ~list_vector_pool()
\-{ dtor
##ADDR## -> 
/-} dtor
END---destruction: named_pooled_thing.o
